L64-python-runtime-proto-parsing.md

```diff
@@ -0,0 +1,99 @@
+
+The wrapper function around the import serves several purposes here. First, it puts the user in control of naming the module (in a manner similar to Javascript), meaning the user never has to concern themself with the confusing `_pb2` suffix. Second, the function provides a wrapping layer through which the library can provide guidance in the case of failed imports.
+
+To be precise, we propose the introduction of three new functions with the following signatures:
+
+```python
+def protos(proto_file: Text,
+           runtime_generated: Optional[bool] = True) -> types.ModuleType:
+    pass
+
+def services(proto_file: Text,
+             runtime_generated: Optional[bool] = True) -> types.ModuleType:
+    pass
+
+def protos_and_services(proto_file: Text,
+                        runtime_generated: Optional[bool] = True) -> Tuple[types.ModuleType, types.ModuleType]:
+    pass
+```
+
+The final function, `protos_and_services` is a simple convenience function allowing the user to import protos and services in a single function call.
+
+The change will be entirely backward compatible. Users manually invoking `protoc` today will not be required to change their build process.
+
+### Import Paths
+
+These functions will behave like normal `import` statements. `sys.path` will be used to search for `.proto` files. The path under which each particular `.proto` file was found will be passed to the protobuf parser as the root of the tree (equivalent to the `-I` flag of `protoc`). This means that a file located at `${SYS_PATH_ENTRY}/foo/bar/baz.proto` will result in the instantation of a module with the fully qualified name `foo.bar.baz_pb2`.  Users are expected to have a directory structure mirroring their desired import structure.
+
+Users have reported that getting Protobuf paths to match Python import paths is quite tricky today using `protoc`. In the case of a failure to import, the library will print troubleshooting information alongside the error.
+
+### Arbitrary Protos
+
+It should be stated that, in practice, it is possible to use these new functions to load totally arbitrary protos. Suppose you wrote a server that took `.proto` files as inputs from clients, instatiated modules from them, and returned some data about the file. For example, the number of message types contained within the supplied file. This could become problematic as new syntax features are added to the Protobuf specification. In the worst case, this would require a redeploy of the server with a sufficiently up-to-date version of `grpcio-tools`. But, regardless, we will claim no support for this use case. The intent of these functions is to enable import of *fixed* `.proto` files, known at build time.
```

Comment:
Should we support runtime reloading? So, the application can invoke `protos_and_services` multiple times during development.

Reply:
I don't think this is a good idea. Suppose you instantiate some messages before reloading and save them off. Then reload the module. Create some more messages. The messages before the reload cannot interact at all with the messages after the reload. You cannot fill one's containers with instances of the others. You cannot equality compare them.

Furthermore, this is already possible in the current design by doing

```python
bar_protos = grpc.protos("foo/bar.proto")
reload(bar_protos)
```

In any case, it's much safer to simply restart the process.

Reply:
Good to know that is possible. As you mentioned, this pattern is not recommended. Then we should not recommend user to restart instead of reload?

Reply:
I'll add a section clarifying that these functions are idempotent and will not cause a reload from disk.